### ISSUE TITLE
Add the ability to use IIAsyncEnumerable<T> as a streams source

### ISSUE THREAD

**Issue by SeanFarrow** [akkadotnet/akka.net#5836]
I'd like to be able to use an IAsyncEnumerable<T> as a source for Akka Streams.

Ideally, this could be used like the Source.From(IEnumerable<T>) stage.

I'm looking to take a method that returns an IAsyncEnumerable<T> and run some stages on it's output.

This is porting some Csv parsing code using a third-party library as the current CSV stage in the Alpakka project doesn't quite give me an equivalent, so for ease of use and quickness, this seems like the easiest option.

**Comment by to11mtm** [comment]
This is napkin code, but if you're just trying to use like `Source.From` have you considered something along the lines of below?

```csharp
    public static class AsyncEnumerableSource
    {
        //Maybe also an overload to create enumerable
        //Could be an ext method as well
        //(Sugar to taste)
        private static Source<T, NotUsed> 
            ViaUnfoldResourceAsync<T>(IAsyncEnumerable<T> e)
        {
            return Source.UnfoldResourceAsync(
                create: () =>
                    Task.FromResult(e.GetAsyncEnumerator()),
                read: async r =>
                {
                    if (await r.MoveNextAsync())
                    {
                        return new Option<T>(r.Current);
                    }
                    else
                    {
                        return Option<T>.None;
                    }
                },
                close: async en =>
                {
                    await en.DisposeAsync();
                    return Done.Instance;
                });
        }
    }
```

That said a 'proper' stage will likely be more performant and I'm all for it.

**Comment by SeanFarrow** [comment]
That would work, but I was ideally looking for a stage, as you say, for performance reasons.
I’ll try and write one, I just need to figure out how to test it.

**Comment by to11mtm** [comment]
@SeanFarrow  may want to look at this: https://stackoverflow.com/a/69344982/2937845 , looks like a good start!

**Comment by Aaronontheweb** [comment]
I thought this was built into Akka.NET a while ago, or am I missing something?

**Comment by Aaronontheweb** [comment]
Ah nevermind - this is using `IAsyncEnumerable` as an input, rather than as an output - correct? I don't believe we support that at the moment but we should definitely add it IMHO.

**Comment by SeanFarrow** [comment]
Yes, exactly, it’s about using IAsyncEnumerable as an input.
I’ll have a look at this over easter, once I can work out how streams stages are tested.

**Comment by ismaelhamed** [comment]
Another options, since there's support for Rx.NET in Akka.Streams, but I don't know how *performant* this is:

IAsyncEnumerable -> ToObservable (Ix.net) -> Source.FromObservable

**Comment by Aaronontheweb** [comment]
Released in v1.4.40
